Implement the API version of the Docker containers interface

natefoo opened the pull request:
Initially when I'd started work on adding Docker Swarm support for GIEs, I was just augmenting the existing (command line) docker calls in the GIE module. Before long, the this grew in to creating a library for interacting with containerization systems. I moved the command line work over to the new library and expanded it, adding a generalized parser for the command line's column output, among other things.

Although this worked at the time, the correct route would've been to use the Docker API, but I fell a bit for the sunk cost fallacy and finished things up using the command line. The command line stuff is now incompatible with at least some commands in the latest docker versions.

Soooooo, I've finally implemented the API versions of the Docker and Docker Swarm interfaces and deprecated the command line interfaces, to be removed in the release after 18.05.

This also includes some general stability improvements for the swarm manager, and it touches the tool dependency docker util slightly, as there was already a model for a docker volume in there. Rather than duplicate, I moved that model over to the containers lib.

natefoo committed:
Rough implentation of the DockerAPIInterface.

natefoo committed:
Create a general method for handling galaxy.containers param ->
docker-py param mapping. Plus some cleanup.

natefoo committed:
Corrections to Docker spec handling

natefoo committed:
WIP Docker Swarm API interface

natefoo committed:
Docker API TLS support

natefoo committed:
Docker service fixes

natefoo committed:
Enable setting defaults for docker API data structures.

natefoo committed:
Docker swarm interface for the API

natefoo committed:
Lint fixes

natefoo committed:
Fixes for swarm manager.

natefoo committed:
Move DockerVolume to containers lib model and fix bugs.

natefoo committed:
More detailed swarm manager state logging

natefoo committed:
Attempt to handle "phantom" docker tasks

natefoo committed:
Fix lint/unit test errors

natefoo committed:
Add some error handling to the Docker API interface

natefoo committed:
Fix lint error

natefoo committed:
Move json format function to galaxy.util.json

natefoo committed:
Add conditional docker dependency for containers interface

jmchilton commented:
This would require adding galaxy.containers to galaxy-lib since that DockerVolume class is used outside the project. I looked over the source and it looks like you are conditionally importing everything that isn't in galaxy-lib or a hard dependency - so that should be fine.

natefoo commented:
Yes, good, my library shall become the standard. ;D

natefoo committed:
Handle the case where a node claims a task exists but the tasks/services
API disagrees

natefoo committed:
Allow docker client callers to pass in a test function that the client
wrapper will call on certain types of expected failures. If the function
returns a truthy value, it is returned instead. This is useful for
state-altering changes where the request may have succeeded but the
response failed.

natefoo committed:
Use consistent whitespace in containers lib classes

natefoo committed:
Containers:
- Improve documentation of docker API param mapping
- Improve sample containers config file
- Fix constraints
- A couple other minor bug/logging fixes

natefoo committed:
Docker: Don't retry 4xx errors indefinitely. Possibly need to do this
for more APIErrors.

natefoo committed:
Infinite loop retry is probably bad. This should help with surviving
transient errors.

natefoo committed:
Fix tests due to missing name in mock module

natefoo committed:
Support manager failover for Docker Swarm mode

natefoo committed:
Swarm manager: log services that do not have tasks on any node.

jmchilton commented on lib/galaxy/containers/docker_swarm_manager.py:
```diff
@@ -19,6 +19,12 @@
+
 try:
     import galaxy   # noqa: F401 this is a test import
 except ImportError:
```
Can you do this without the sys.path edit stuff? I don't think importing things from galaxy-lib should be modifying the path like this - the whole library should be importable without side effects.

Can you keep this script clean, even define a main, but then place something in ``scripts/`` that messes with the sys.path and then just calls out the main here? Like this (https://github.com/galaxyproject/galaxy/blob/dev/scripts/create_db.py) maybe?

natefoo committed:
Remove swarm manager from lib

natefoo commented:
@jmchilton how about e651b34156a3af022bac250464d92ca6394c3d85?

Unrelated, back in WIP due to the DockerAPIClient class still not working as intended: failover doesn't work properly because bound APIClient instancemethods are passed around in the handler staticmethod, so making new APIClients in DockerAPIClient does nothing for a single call. A class var with the APIClient instance and then passing the method name to the handler rather than the method itself might work.

natefoo committed:
Load containers interface on app load if enable_beta_containers_interface is set.

natefoo committed:
Fix docker interface failover by making the docker.APIClient wrapper use
class vars, since we only ever want callers to use one APIClient.

natefoo committed:
Make the swarm manager recover on uncaught exceptions. It should do a
better job of catching them in the application, but this is good enough
for the moment.

natefoo committed:
Fix lint error

jmchilton commented:
> @jmchilton how about e651b34?

@natefoo If you prefer that approach that is fine with me. Seems like the script might be useful as a stand-alone thing installable through pip if we placed it cleanly in galaxy-lib but if that isn't a concern for you I'm fine.

natefoo commented:
I think it's useful, but probably nobody but me is going to use it...

I really am surprised something doesn't already exist for this purpose and I expect something better to come along at some point.
